pkg/capabilities/consensus/ocr3_store.go

```diff
@@ -0,0 +1,168 @@
+package consensus
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+)
+
+type store struct {
+	requestIDs []string
+	requests   map[string]*request
+
+	requestExpiredAfter time.Duration
+	now                 func() time.Time
+
+	mu sync.RWMutex
+}
+
+func newStore(requestExpiredAfter time.Duration, now func() time.Time) *store {
+	return &store{
+		requestIDs:          []string{},
+		requests:            map[string]*request{},
+		requestExpiredAfter: requestExpiredAfter,
+		now:                 now,
+	}
+}
+
+func (s *store) add(ctx context.Context, req *request) ([]*request, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	now := s.now()
+	req.ExpiresAt = now.Add(s.requestExpiredAfter)
+	s.requestIDs = append(s.requestIDs, req.RequestID)
+	s.requests[req.RequestID] = req
+
+	return s.evictExpiredBefore(ctx, now)
+}
+
+func (s *store) get(ctx context.Context, requestID string) (*request, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	r, ok := s.requests[requestID]
+	if !ok {
+		return nil, fmt.Errorf("request with id %s not found", requestID)
+	}
+
+	return r, nil
+}
+
+func (s *store) firstN(ctx context.Context, batchSize int) ([]*request, error) {
+	if batchSize == 0 {
+		return nil, errors.New("batchsize cannot be 0")
+	}
+	if len(s.requestIDs) < 1 {
+		return nil, errors.New("queue is empty")
+	}
+
+	got := []*request{}
+	newRequestIDs := []string{}
+	lastIdx := 0
+	for idx, r := range s.requestIDs {
+		gr, ok := s.requests[r]
+		if !ok {
+			continue
+		}
+
+		got = append(got, gr)
+		newRequestIDs = append(newRequestIDs, r)
+		lastIdx = idx
+		if len(got) == batchSize {
+			break
+		}
+	}
+
+	s.requestIDs = append(newRequestIDs, s.requestIDs[lastIdx+1:]...)
+	if len(got) == 0 {
+		return nil, errors.New("queue is empty")
+	}
+	return got, nil
+}
+
+func (s *store) evict(ctx context.Context, requestID string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	_, ok := s.requests[requestID]
+	if !ok {
+		return fmt.Errorf("request with id %s not found", requestID)
+	}
+
+	delete(s.requests, requestID)
+	return nil
+}
+
+func (s *store) evictExpiredBefore(ctx context.Context, now time.Time) ([]*request, error) {
+	idx, err := binarySearch(s.requests, s.requestIDs, 0, len(s.requestIDs)-1, now)
```

Comment:
There's no point in running binary search if you later iterate over the first segment of requestIDs array anyway. Just add a timestamp check in the loop below instead of index comparison.

Reply:
Wow... 🤦 I think I lost some perspective when writing this PR! Thanks for catching!